File: apps/api_py/lib/crypto.py
import os
import binascii
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
import config as cfg


def _get_key() -> bytes:
    key = binascii.unhexlify(cfg.MASTER_KEY)
    if len(key) != 32:
        raise ValueError(
            "CPWORK_MASTER_KEY must be 32 bytes hex (64 hex chars). "
            "Generate with: openssl rand -hex 32"
        )
    return key


def encrypt_secret(plaintext: str) -> str:
    """Encrypt plaintext. Returns iv:tag:ciphertext (hex)."""
    iv = os.urandom(12)
    aesgcm = AESGCM(_get_key())
    ciphertext_with_tag = aesgcm.encrypt(iv, plaintext.encode("utf-8"), None)
    # AESGCM.encrypt appends the 16-byte tag at the end
    ciphertext = ciphertext_with_tag[:-16]
    tag = ciphertext_with_tag[-16:]
    return ":".join([iv.hex(), tag.hex(), ciphertext.hex()])


def decrypt_secret(payload: str) -> str | None:
    """Decrypt a value produced by encrypt_secret. Returns None on failure."""
    try:
        parts = payload.split(":")
        if len(parts) != 3:
            return None
        iv_hex, tag_hex, data_hex = parts
        if not iv_hex or not tag_hex or not data_hex:
            return None
        iv = bytes.fromhex(iv_hex)
        tag = bytes.fromhex(tag_hex)
        ciphertext = bytes.fromhex(data_hex)
        aesgcm = AESGCM(_get_key())
        plaintext = aesgcm.decrypt(iv, ciphertext + tag, None)
        return plaintext.decode("utf-8")
    except Exception:
        return None

File: apps/api/src/lib/crypto.ts
import crypto from 'node:crypto';
import { config } from '../config.js';

const ALGO = 'aes-256-gcm';

function getKey(): Buffer {
  const key = Buffer.from(config.masterKey, 'hex');
  if (key.length !== 32) {
    throw new Error(
      'CPWORK_MASTER_KEY must be 32 bytes hex (64 hex chars). Generate with: openssl rand -hex 32',
    );
  }
  return key;
}

/** Encrypt a plaintext secret. Returns iv:tag:ciphertext (hex). */
export function encryptSecret(plaintext: string): string {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv(ALGO, getKey(), iv);
  const enc = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  return [iv.toString('hex'), tag.toString('hex'), enc.toString('hex')].join(':');
}

/** Decrypt a value produced by encryptSecret. Returns null on failure. */
export function decryptSecret(payload: string): string | null {
  try {
    const [ivHex, tagHex, dataHex] = payload.split(':');
    if (!ivHex || !tagHex || !dataHex) return null;
    const decipher = crypto.createDecipheriv(
      ALGO,
      getKey(),
      Buffer.from(ivHex, 'hex'),
    );
    decipher.setAuthTag(Buffer.from(tagHex, 'hex'));
    const dec = Buffer.concat([
      decipher.update(Buffer.from(dataHex, 'hex')),
      decipher.final(),
    ]);
    return dec.toString('utf8');
  } catch {
    return null;
  }
}
